metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,643 @@
+        Constant.DATASET_LIST: "{POWERBI_ADMIN_BASE_URL}/groups/{WORKSPACE_ID}/datasets",
+    }
+
+    def create_scan_job(self, workspace_id: str) -> str:
+        """
+        Create scan job on PowerBI for the workspace
+        """
+        request_body = {"workspaces": [workspace_id]}
+
+        scan_create_endpoint = AdminAPIResolver.API_ENDPOINTS[Constant.SCAN_CREATE]
+        scan_create_endpoint = scan_create_endpoint.format(
+            POWERBI_ADMIN_BASE_URL=DataResolverBase.ADMIN_BASE_URL
+        )
+
+        logger.debug("Creating metadata scan job, request body (%s)", request_body)
+
+        res = requests.post(
+            scan_create_endpoint,
+            data=request_body,
+            params={
+                "datasetExpressions": True,
+                "datasetSchema": True,
+                "datasourceDetails": True,
+                "getArtifactUsers": True,
+                "lineage": True,
+            },
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+
+        res.raise_for_status()
+        # Return scan_id of Scan created for the given workspace
+        scan_id = res.json()["id"]
+
+        logger.info("Scan id({})".format(scan_id))
+
+        return scan_id
+
+    @staticmethod
+    def _calculate_max_trial(minimum_sleep: int, timeout: int) -> int:
+        if timeout < minimum_sleep:
+            logger.info(
+                f"Setting timeout to minimum_sleep time {minimum_sleep} seconds"
+            )
+            timeout = minimum_sleep
+
+        return timeout // minimum_sleep
+
+    def wait_for_scan_to_complete(self, scan_id: str, timeout: int) -> Any:
+        """
+        Poll the PowerBi service for workspace scan to complete
+        """
+        minimum_sleep = 3
+        max_trial: int = AdminAPIResolver._calculate_max_trial(minimum_sleep, timeout)
+        logger.info(f"Max trial {max_trial}")
+
+        scan_get_endpoint = AdminAPIResolver.API_ENDPOINTS[Constant.SCAN_GET]
+        scan_get_endpoint = scan_get_endpoint.format(
+            POWERBI_ADMIN_BASE_URL=DataResolverBase.ADMIN_BASE_URL, SCAN_ID=scan_id
+        )
+        logger.debug(f"Hitting URL={scan_get_endpoint}")
+        trail = 1
+        while True:
+            logger.info(f"Trial = {trail}")
+            res = requests.get(
+                scan_get_endpoint,
+                headers={Constant.Authorization: self.get_access_token()},
+            )
+            if res.status_code != 200:
+                message = f"API({scan_get_endpoint}) return error code {res.status_code} for scan id({scan_id})"
+                logger.warning(message)
+                raise ConnectionError(message)
+
+            if res.json()["status"].upper() == "Succeeded".upper():
+                logger.info(f"Scan result is available for scan id({scan_id})")
+                return True
+
+            if trail == max_trial:
+                break
```

Comment:
I think we should log something about maximum retry reached

Reply:
done